### ISSUE TITLE
 SNN is able to recognize correctly only the first element passed at inference time

### ISSUE THREAD

**Issue by NicolaCST** [fangwei123456/spikingjelly#323]
**Issue type**

- [ ] Bug Report
- [ ] Feature Request
- [x] Help wanted
- [ ] Other

**SpikingJelly version**

`0.0.0.0.13` - latest

**Description**
Hi, 
I'm experiencing some weird behaviour during net's inference.  I'll try to be as clear as possible:
I can successfully train a standard SNN on MNIST

(Epoch 100:
B train acc: 0.9198 -- B train Loss: 0.0154 ## B valid acc: 0.9151 -- B valid loss: 0.0160  ## Time: 9.44 )

However, there are some problems at inference time. 

# It seems like that the net is able to recognize correctly only the first element passed at inference time.
Es.: train a net -> inference asking for test_data[2] -> correct output -> ask any other test_data[n] -> No response ( Firing rate: [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]] )  (no matter the number of times i pass the same image)

I can't figure out what could be the problem... thanks in advance for the help
@fangwei123456 

**Minimal code to reproduce the error/bug**
```python


from spikingjelly.activation_based import neuron, encoding, functional, surrogate, layer
from spikingjelly import visualizing
import torch
import torch.nn as nn
import torchvision
import numpy as np
from torch.utils.tensorboard import SummaryWriter
from torch.cuda import amp
import torch.nn.functional as F


# SNN net class
#automatic switch between single-step / multi-step based on passed param T
class SNN(nn.Module):
  def __init__(self, **kwargs):
      super().__init__()

      self.layer = nn.Sequential(
          layer.Flatten(),
          layer.Linear(28 * 28, 10, bias=False),
          neuron.IFNode(surrogate_function=surrogate.ATan()))
      
      if(len(kwargs) > 0):
        self.T = kwargs['T']
        functional.set_step_mode(self, step_mode="m")
        functional.set_backend(self, backend="cupy")

  def forward(self, x:torch.Tensor):
    if(hasattr(self, 'T')):
      x_seq = x.unsqueeze(0).repeat(self.T, 1, 1, 1, 1)
      x_seq = self.layer(x_seq)
      out = x_seq.mean(0)
    else:
      out = self.layer(x)
    return out


#Train and Eval
# returns a dict with train accuracy - loss (and validation, if true) + time
def fit_and_eval(net, epochs, train_data_loader, validation_data_loader, device,  T=5, validation_step = False):
  dict_hist = {
      'Train_acc' : [],
      'Train_loss' : [],
      'Valid_acc' : [],
      'Valid_loss': [],
      'Time_train': []
  }
  
  # Gradient scaling helps prevent gradients with small magnitudes from 
  # flushing to zero (“underflowing”) when training with mixed precision.
  scaler = amp.GradScaler()
  
  functional.reset_net(net)

  for epoch in range(epochs):
      print("Epoch {}:".format(epoch+1))

      # ------------- TRAIN SET ----------- # 
      net.train()
      train_loss = 0
      train_acc = 0
      train_samples = 0

      time_train_batch = timing.time()

      for img, label in train_data_loader:
          optimizer.zero_grad()
          img = img.to(device)
          label = label.to(device)
          label_onehot = F.one_hot(label, 10).float()

          with amp.autocast():
            if(hasattr(net, 'T') == False):
              out_fr = 0.
              for t in range(T):
                  encoded_img = encoder(img) 
                  out_fr += net(encoded_img)
              out_fr = out_fr / T   
            else:
              encoded_img = encoder(img)
              out_fr = net(encoded_img)

            loss = F.mse_loss(out_fr, label_onehot)
            scaler.scale(loss).backward()
            scaler.step(optimizer)
            scaler.update()

          
          train_samples += label.numel()
          train_loss += loss.item() * label.numel()
          train_acc += (out_fr.argmax(1) == label).float().sum().item()

          functional.reset_net(net)

      time_end = timing.time()

      train_loss /= train_samples
      train_acc /= train_samples

      dict_hist['Train_acc'].append(train_acc)
      dict_hist['Train_loss'].append(train_loss)
      dict_hist['Time_train'].append(time_end - time_train_batch)

      # ------------- VALIDATION ----------- # 
      if(validation_step == True):
        net.eval()
        val_loss = 0
        val_acc = 0
        val_samples = 0

        with torch.no_grad():
            for img, label in validation_data_loader:
                img = img.to(device)
                label = label.to(device)
                label_onehot = F.one_hot(label, 10).float()

                if(hasattr(net, 'T') == False):
                  out_fr = 0.
                  for t in range(T):
                      encoded_img = encoder(img)
                      out_fr += net(encoded_img)
                  out_fr = out_fr / T
                  loss = F.mse_loss(out_fr, label_onehot)
                else:
                  encoded_img = encoder(img)
                  out_fr = net(encoded_img)
                  loss = F.mse_loss(out_fr, label_onehot)

                val_samples += label.numel()
                val_loss += loss.item() * label.numel()
                val_acc += (out_fr.argmax(1) == label).float().sum().item()

                functional.reset_net(net)

        val_loss /= val_samples
        val_acc /= val_samples

        dict_hist['Valid_acc'].append(val_acc)
        dict_hist['Valid_loss'].append(val_loss)

      if(validation_step == True):
        print("B train acc: {:.4f} -- B train Loss: {:.4f} ## B valid acc: {:.4f} -- B valid loss: {:.4f}  ## Time: {:.2f}".format(train_acc, train_loss, val_acc, val_loss, (time_end - time_train_batch)))
      else:
        print("B train acc: {:.4f} -- B train Loss: {:.4f} ## Time: {:.2f}".format(train_acc, train_loss, (time_end - time_train_batch)))

  return dict_hist


#Data Loaders
def init_train_loader(batch, shuffle:bool=True):
  train_data_loader = data.DataLoader(
          dataset = train_dataset,
          batch_size = batch,
          shuffle = shuffle,
          drop_last = True)  
  return train_data_loader


def init_test_loader(batch, shuffle:bool=False): 
  test_data_loader = data.DataLoader(
          dataset = test_dataset,
          batch_size = batch,
          shuffle = shuffle,
          drop_last = False)
  return test_data_loader


net = SNN()
net.to(device)

epochs = 100
batch_size = 64
lr = 1e-3

# Use Adam optimizer
optimizer = torch.optim.Adam(net.parameters(), lr=lr)

# Use Poisson encoder
encoder = encoding.PoissonEncoder()

# Init data loaders
train_data_loader = init_train_loader(batch_size)
test_data_loader = init_test_loader(batch_size)

train_val_ls = fit_and_eval(net, epochs, train_data_loader, test_data_loader, device, writer, T=50, validation_step=True)
# Epoch 100:
# B train acc: 0.9198 -- B train Loss: 0.0154 ## B valid acc: 0.9151 -- B valid loss: 0.0160  ## Time: 9.44



################# INFERENCE #####################

net.eval()
# 注册钩子
output_layer = net.layer[-1] # 输出层
output_layer.v_seq = []
output_layer.s_seq = []

def save_hook(m, x, y):
    m.v_seq.append(m.v.unsqueeze(0))
    m.s_seq.append(y.unsqueeze(0))

output_layer.register_forward_hook(save_hook)

T = 50 #tested also 20, 500, 100 

with torch.no_grad():
    img, label = test_dataset[4] #Here, the first value is correct, if I change this value, there are no responses
    img = img.to(device)
    out_fr = 0.
    for t in range(T): 
      encoded_img = encoder(img)
      out_fr += net(encoded_img)
    out_spikes_counter_frequency = (out_fr / T).cpu().numpy()
    print(f'Firing rate: {out_spikes_counter_frequency}')

    output_layer.v_seq = torch.cat(output_layer.v_seq)
    output_layer.s_seq = torch.cat(output_layer.s_seq)
    v_t_array = output_layer.v_seq.cpu().numpy().squeeze()  # v_t_array[i][j]表示神经元i在j时刻的电压值

    s_t_array = output_layer.s_seq.cpu().numpy().squeeze()  # s_t_array[i][j]表示神经元i在j时刻释放的脉冲，为0或1

```

**Comment by Lyu6PosHao** [comment]
maybe _functional.reset_net(net)_ is needed after each time of inference? @NicolaCST

**Comment by NicolaCST** [comment]
Ye after some research with my colleague we found out the problem, and was on the hook function + storing part. Now we are able to correctly visualize the spikes. However @Lyu6PosHao that part was also needed,  thanks!

```python

net.eval()

output_layer = net.conv_fc[-1]
output_layer.store_v_seq = True

visualisation = []

def hook_fn(m, i, o):
  visualisation.append(o)

output_layer.register_forward_hook(hook_fn)

with torch.no_grad():
    img, label = test_dataset[2]
    img = img.to(device)

    encoded_img = encoder(img)
    out_fr = net(encoded_img)

    print(f'Firing rate: {out_fr}')

v_t_array = output_layer.v_seq.squeeze().cpu().numpy()
s_array = visualisation[0].squeeze().clone().detach()

```
